### PR TITLE
GRRafana HTTP server documentation

### PR DESCRIPTION
**Pull request by tsehori**
Added documentation about installing and using the new GRRafana HTTP server. Also restructured the documentation page on Monitoring a little bit, by aggregating all example setups under the same sub-heading for better organization.

Note that this PR is relevant only once google/grr#882 is merged, as it contains all the changes discussed here.

### CODE REVIEW

File: maintaining-and-tuning/monitoring.md [hunk 1]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+component of GRR server (change the port according to the GRR server
+component you want) and Fleetspeak servers, and if you'd like to
+create your own
+[custom dashboards](https://grafana.com/docs/grafana/latest/getting-started/getting-started/#create-a-dashboard),
+[panels](https://grafana.com/docs/grafana/latest/panels/add-a-panel/#add-a-panel)
+and [alerts](https://grafana.com/docs/grafana/latest/alerting/create-alerts/#create-alerts),
+make sure to go over the corresponding documentation in Grafana.
 
-1. [Set up Prometheus as a data source](https://grafana.com/docs/grafana/latest/features/datasources/prometheus/#prometheus-data-source) for Grafana, so that Grafana can display the metrics from Prometheus. To do that, follow the guide [here](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/#add-a-data-source) (in step 3 of the guide, we suggest naming the data source `grr-server` so that it will match the configuration file for the provided sample dashboards later on. If it is a Fleetspeak-enabled GRR deployment, add a second datasource named `fleetspeak`).
+## Real-World Setup
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:273]: I'm not sure what a "Real World Setup" header means. What is this section about?

File: maintaining-and-tuning/monitoring.md [hunk 2]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+service exports read-only metrics under `/metrics` and `/varz` and does
+**not enforce any access control**. People with access to it can read
+aggregated metrics about your GRR installation. With these metrics, facts
+about the number of workers, flow activity, and service health can be
+derived. Make sure to limit access to the port, for example by employing
+a firewall. Furthermore, read [Prometheus Security](https://prometheus.io/docs/operating/security/).
+
+## Monitoring Client Load Stats
+
+In GRR Admin UI you are able to view statistics of an individual GRR
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:305]: We plan to deprecate looking at the client statistics from the UI, so probably makes sense to not mention it here.

File: maintaining-and-tuning/monitoring.md [hunk 3]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+client. To do this, enter the host information of the client, and on
+the menu, click Advanced and then Client Load Stats. This data is
+gathered from the GRR client itself.
+If you use Grafana dashboards to monitor other GRR server components,
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:309]: Let's describe this as *the* way to graph client load stats, not as a "backup" option.

File: maintaining-and-tuning/monitoring.md [hunk 4]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+correctly) and specifiy its
+[type](https://grafana.com/docs/grafana/latest/variables/variable-types/#variables-types)
+as "Constant" (and make sure to not hide the variable or the label,
+so that it's visible in the dashboard). At this point, you should
+have an empty dashboard with an empty textbox named `ClientID` above
+your panels. Feel free to create new panels and playing around with
+all the individual statistics information** you can get from your
+clients!
+
+1. Similarly to before, if you do not want to create your own Grafana monitoring
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:354]: I would mention this first and then say that the dashboard might be extended.

File: maintaining-and-tuning/monitoring.md [hunk 5]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+dashboards for individual client monitoring, we got you. In the
+[sample dashboards in step 3](#example-visualization-and-alerting-setup), you
+can also find a sample dashboard called Client Load Stats for this exact
+purpose.
+
+** **Note:** not all the metrics gathered by Grafana are related to individual
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:360]: I think we should only add this section when we have a sample dashboard ready. Otherwise it's too cumbersome. Wdyt?

**Review by mbushkov — CHANGES_REQUESTED** [review]
Looks good! Please see my comments below.

File: maintaining-and-tuning/monitoring.md [hunk 6]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
-1. Grafana is now set up and can be visited at `http://localhost:3000`. The username and password should be "admin"; please change it.
-If the Grafana UI doesn't show up, either the [installation of Grafana](https://grafana.com/docs/grafana/latest/installation/#install-grafana) or the [server run](https://grafana.com/docs/grafana/latest/installation/debian/#2-start-the-server) failed; make sure to check the official documentation.
+1. Grafana is now set up and can be visited at `http://localhost:3000`.
+The username and password should be "admin"; please change it.
+If the Grafana UI doesn't show up, either the
+[installation of Grafana](https://grafana.com/docs/grafana/latest/installation/#install-grafana)
+or the [server run](https://grafana.com/docs/grafana/latest/installation/debian/#2-start-the-server)
+failed; make sure to check the official documentation.
+
+1. [Set up Prometheus as a data source](https://grafana.com/docs/grafana/latest/features/datasources/prometheus/#prometheus-data-source)
```
[the remark this answers is not in the document]
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:209]: I think your'e right that we don't want to refer too much to other guides. On the other hand, Grafana UI might change from time to time (IIRC it did at least once in the past few months), and so referring to the official Grafana documentation will keep GRR's documentation consistent and correct. This applies to screenshots as well. Let me know what you think.

Essentially, given the default GRR setup, Grafana will have to still be installed by the user (as we don't want to include it in GRR's deb package) and the data source has to be configured as well.

File: maintaining-and-tuning/monitoring.md [hunk 7]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+into Grafana.
+You can choose to import sample dashboards from the
+[Grafana folder in GRR repository](https://github.com/google/grr/monitoring/grafana)
+before creating your own. These dashboards contain some example graphs of
+metrics scraped by Prometheus, and also implement sample alerts. To do that,
+first download the dashboards from the repository, and then head over to
+`http://localhost:3000/dashboard/import`. There, you can click 'Upload
+.json file' and upload the dashboard you have downloaded from the repository.
+The dashboard is now imported; you can access it by going to
+`http://localhost:3000/dashboards` and clicking the dashboard's name, e.g
```
[the remark this answers is not in the document]
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:231]: I think that the current folder structure is very expressive, as the dashboards of both deployments [are separated](https://github.com/google/grr/tree/master/monitoring/grafana/grr_grafanalib_dashboards); but I do agree that we can be more precise. I'll elaborate on this.

File: maintaining-and-tuning/monitoring.md (hunk 1 above)
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:273]: This section provides brief information on service discovery in Prometheus and the security considerations. Do you think it should be elaborated or removed?

File: maintaining-and-tuning/monitoring.md (hunk 5 above)
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:363]: I agree. To add on top of that, the metrics' names are quite readable and are differentiable from one another, so I don't see a reason to confuse our users with this note at this time.

File: maintaining-and-tuning/monitoring.md [hunk 8]
```diff
@@ -198,16 +199,189 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+panels; each such
+[panel](https://grafana.com/docs/grafana/latest/panels/panels-overview/#panel-overview)
+consists of a graph that contain one or more metrics queried from Prometheus,
+and possibly alerting rules.
+
+1. If you want to use the alerts, you first need to define a
+[notification channel](https://grafana.com/docs/grafana/latest/alerting/notifications/#alert-notifications)
+for the alerts. This can be done by heading over to
+`http://localhost:3000/alerting/notification/new` and
+[following the form](https://grafana.com/docs/grafana/latest/alerting/notifications/#new-notification-channel-fields)
```
[the remark this answers is not in the document]
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:278]: I added a screenshot; let me know if that's what you had in mind.

**Comment by tsehori** [comment]
@mbushkov I addressed most of your comments and left some of my own. Let me know what you think :)
PTAL

File: maintaining-and-tuning/monitoring.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mbushkov) [reply maintaining-and-tuning/monitoring.md:239]: I think we should provide links to Grafana documentation, but also effectively list the commands that the reader can copy-paste into the terminal to get things going at least for the simple default setup (i.e. when GRR is installed from the release server deb). It's true that the instructions might get out of date eventually - in that case we'd need to update them. Giving that Grafana is pretty stable and that there's a value in having a "all you need to know is in this doc" kind of manual, I think that providing detailed instructions is worth it.

File: maintaining-and-tuning/monitoring.md [hunk 9]
```diff
@@ -162,15 +162,45 @@ threadpool_threads{job="grr_worker"}
 ```	
 rate(grr_client_crashes_total{job="grr_worker"}[5m])
 ```
-### Example visualization and alerting setup
+
+## Real-World Setup
+Although Prometheus can display basic charts using the
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:167]: I think we should describe Grafana not as an option but as something that a user wants to use to display Prometheus graphs. I.e. at the top of the page we can put something: "This document describes how to set up Prometheus and Grafana to monitor GRR server components."

File: maintaining-and-tuning/monitoring.md [hunk 10]
```diff
@@ -162,15 +162,45 @@ threadpool_threads{job="grr_worker"}
 ```	
 rate(grr_client_crashes_total{job="grr_worker"}[5m])
 ```
-### Example visualization and alerting setup
+
+## Real-World Setup
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:166]: "Real-World" is a confusing term here. Does the rest of the doc describe something other than the real world setup? :) I understand that the intention here is to talk about security and implications of using Prometheus/Grafana in larger setups. But then I'd structure the whole document as:
1. Here's how to setup Prometheus.
1. Prometheus security (and maybe scaling).
1. Here's how to setup Grafana.
1. Grafana security.

**Review by mbushkov — CHANGES_REQUESTED** [review]

**Comment by tsehori** [comment]
I restructured the document a little bit to be more consistent and clear.
I also added some missing instructions while referencing to the official documentation of Prometheus and Grafana.
Let me know what you think! PTAL

File: maintaining-and-tuning/monitoring.md [hunk 11]
```diff
@@ -198,16 +224,160 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+or the [server run](https://grafana.com/docs/grafana/latest/installation/debian/#2-start-the-server)
+failed; make sure to check the official documentation.
+
+1. [Set up Prometheus as a data source](https://grafana.com/docs/grafana/latest/features/datasources/prometheus/#prometheus-data-source)
+for Grafana, so that Grafana can display the metrics from Prometheus.
+To do that, follow the
+[these instructions](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/#add-a-data-source):
+Click on Configuration -> Data Sources in the Grafana server UI, then click
+on the "Add data source" button. From the list of "time series databases", choose
+Prometheus and in the Name field type `grr-server` (for Fleetspeak, type `fleetspeak`).
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:240]: Should both data sources be added for FS-based deployments?

File: maintaining-and-tuning/monitoring.md [hunk 12]
```diff
@@ -198,16 +224,160 @@ These instructions assume that GRR server and Prometheus are both up and running
...
+
+1. Grafana is set up and ready to show metrics scraped by Prometheus. You
+can start by either
+[creating your own dashboards](https://grafana.com/docs/grafana/latest/getting-started/getting-started/#create-a-dashboard)
+or [importing exisiting dashboards](https://grafana.com/docs/grafana/latest/reference/export_import/#importing-a-dashboard)
+into Grafana.
+You can choose to import sample dashboards from the
```
Comment (mbushkov) [inline maintaining-and-tuning/monitoring.md:247]: I'd rephrase it along the lines of: "The next step would be to import dashboards provided with GRR. You can customize and expand them with additional graphs, if needed".
Basically, we have to present the shipped dashboards as a main option.

**Review by mbushkov — COMMENTED** [review]
Looks really good, please see my 2 comments below.

File: maintaining-and-tuning/monitoring.md (hunk 11 above)
Comment (tsehori) [reply maintaining-and-tuning/monitoring.md:240]: The [sample dashboards](https://github.com/google/grr/tree/master/monitoring/grafana/grr_grafanalib_dashboards) are designed in a way that Fleetspeak can be monitored independently from GRR. If that was the case, then we'd need two separate data sources for GRR and Fleetspeak separately.
In retrospect, that isn't what we aim for. I'll change the documentation here to refer to only one data source `grr-server` and change the [sample dashboard for Fleetspeak](https://github.com/google/grr/blob/fb203630e3e90a40166e9529be8d74dcfbfec76d/monitoring/grafana/grr_grafanalib_dashboards/fleetspeak_enabled_setup/dashboards_to_generate/fleetspeak.dashboard.py#L263) accordingly to reflect this.

Comment (tsehori) [reply maintaining-and-tuning/monitoring.md]: google/grr#896 references this.
